File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void dfs(TreeNode root,int t,List<Integer> cur,List<List<Integer>> res)
    {
        if(root==null) return;
        cur.add(root.val);
        t-=root.val;
        if(t==0&&root.left==null&&root.right==null)
        {
            res.add(new ArrayList<>(cur));
        }   
        else
        {
            dfs(root.left,t,cur,res);
            dfs(root.right,t,cur,res);
        } 
        cur.remove(cur.size()-1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer> cur =new ArrayList<>();
        List<List<Integer>> res=new ArrayList<>();
        dfs(root,targetSum,cur,res);
        return res;
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traverseSum(TreeNode* node, int targetSum, vector<int>& path, vector<vector<int>>& result){
        if(!node) return;

        path.push_back(node->val);
        targetSum -= (node->val);

        if(!node->left && !node->right && targetSum == 0) result.push_back(path);

        traverseSum(node->left, targetSum, path, result);
        traverseSum(node->right, targetSum, path, result);

        path.pop_back();
    }

    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> result;
        vector<int> path;
        traverseSum(root, targetSum, path, result);
        return result;
    }
};